### PR TITLE
Remove mesh reference tracking by name

### PR DESCRIPTION
**Pull request by karl-**
### Purpose of this PR

Fixes an issue where ProBuilder would modify `Mesh.name` on every scene reload. Meshes are now named once at time of creation, and this name is not used to track whether a mesh asset is related to a `ProBuilderMesh` component.

Previously, mesh name was used a means of detecting when a `ProBuilder` mesh was duplicated. It is necessary to create a new `Mesh` asset when this occurs.

This change introduces a non-serialized mesh version tracking index, which allows us to identify when a mesh has been modified vs. copied. I.e., when a mesh is duplicated the serialized tracking index will be non-zero, where-as the non-serialized index will be uninitialized. Through this check we can reliably detect a pasted mesh when listening for new object creation from `ObjectChangeEvents`.

New unit tests are added for copy/paste, duplicate, and prefab changes.

### Links

**Fogbugz:**
**Jira:** STO-2878

### Comments to Reviewers

[List known issues, planned work, provide any extra context for your code.]

### CODE REVIEW

File: Debug/Editor/ResourceExplorer.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace UnityEditor.ProBuilder.Debug
+{
```
Comment (karl-) [inline Debug/Editor/ResourceExplorer.cs:6]: note to reviewers - this is in `Debug/` and not included in package releases

**Review by lopezt-unity — COMMENTED** [review]
It seems that the branch is not up-to-date with master that is on 5.0.7 (changelog here is 5.0.6)

File: Runtime/Core/ProBuilderMeshFunction.cs [hunk 2]
```diff
@@ -377,12 +388,14 @@ public void ToMesh(MeshTopology preferredTopology = MeshTopology.Triangles)
...
+        /// the mesh asset will reference the original instance. If you are making a copy to edit, you must call
+        /// MakeUnique to avoid modifying a shared mesh asset.
         /// </summary>
-        internal void MakeUnique()
+        public void MakeUnique()
         {
-            // set a new UnityEngine.Mesh instance
-            mesh = new Mesh();
+            mesh = new Mesh() { name = $"pb_Mesh{GetInstanceID()}" };
+            // Debug.Log($"<color=red>mesh alloc <b>MakeUnique</b> {name} {GetInstanceID()}</color>");
```
Comment (lopezt-unity) [inline Runtime/Core/ProBuilderMeshFunction.cs:398]: Needs to be removed

**Review by lopezt-unity — APPROVED** [review]
LGTM overall, nice improvement, I'll give it a try this afternoon as well in a test project.

File: Editor/EditorCore/HierarchyListener.cs [hunk 3]
```diff
@@ -1,28 +1,88 @@
...
+        }
+
+        static void ObjectEventChangesPublished(ref ObjectChangeEventStream stream)
+        {
+            for (int i = 0, c = stream.length; i < c; ++i)
+            {
+                // ProBuilderMesh was created via duplicate, copy paste
+                if (stream.GetEventType(i) == ObjectChangeKind.CreateGameObjectHierarchy)
+                {
+                    stream.GetCreateGameObjectHierarchyEvent(i, out CreateGameObjectHierarchyEventArgs data);
```
[the remark this answers is not in the document]
Comment (karl-) [reply Editor/EditorCore/HierarchyListener.cs:33]: the type isn't explicit, so i'd prefer to keep the full declaration here

File: Editor/EditorCore/ProBuilderMeshEditor.cs [hunk 4]
```diff
@@ -124,6 +124,10 @@ public override void OnInspectorGUI()
             MeshFilter filter;
             Mesh sharedMesh = null;
 
+            EditorGUILayout.LabelField("Mesh Sync State", m_Mesh.meshSyncState.ToString());
+            EditorGUILayout.LabelField("m_MeshVersionIndex", m_Mesh.versionIndex.ToString());
+            EditorGUILayout.LabelField("m_NonSerializedMeshIndex", m_Mesh.nonSerializedVersionIndex.ToString());
+
```
[the remark this answers is not in the document]
Comment (karl-) [reply Editor/EditorCore/ProBuilderMeshEditor.cs:130]: we do not

**Pull request merged**